modules/auth/sso/basic.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2014 The Gogs Authors. All rights reserved.
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package sso
+
+import (
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"gitea.com/macaron/macaron"
+	"gitea.com/macaron/session"
+)
+
+// Ensure the struct implements the interface.
+var (
+	_ SingleSignOn = &Basic{}
+)
+
+// Basic implements the SingleSignOn interface and authenticates requests (API requests
+// only) by looking for Basic authentication data or "x-oauth-basic" token in the "Authorization"
+// header.
+type Basic struct {
+}
+
+// Init does nothing as the Basic implementation does not need to allocate any resources
+func (b *Basic) Init() error {
+	return nil
+}
+
+// Free does nothing as the Basic implementation does not have to release any resources
+func (b *Basic) Free() error {
+	return nil
+}
+
+// IsEnabled returns true as this plugin is enabled by default and its not possible to disable
+// it from settings.
+func (b *Basic) IsEnabled() bool {
+	return setting.Service.EnableBasicAuth
+}
+
+// Priority determines the order in which authentication methods are executed.
+// The lower the priority, the sooner the plugin is executed.
+func (b *Basic) Priority() int {
+	return 40000
+}
+
+// VerifyAuthData extracts and validates Basic data (username and password/token) from the
+// "Authorization" header of the request and returns the corresponding user object for that
+// name/token on successful validation.
+// Returns nil if header is empty or validation fails.
+func (b *Basic) VerifyAuthData(ctx *macaron.Context, sess session.Store) *models.User {
+	baHead := ctx.Req.Header.Get("Authorization")
+	if len(baHead) == 0 {
+		return nil
+	}
+
+	auths := strings.Fields(baHead)
+	if len(auths) != 2 || auths[0] != "Basic" {
+		return nil
+	}
+
+	var u *models.User
+	uname, passwd, _ := base.BasicAuthDecode(auths[1])
+
+	// Check if username or password is a token
+	isUsernameToken := len(passwd) == 0 || passwd == "x-oauth-basic"
+	// Assume username is token
+	authToken := uname
+	if !isUsernameToken {
+		// Assume password is token
+		authToken = passwd
+	}
+
+	uid := CheckOAuthAccessToken(authToken)
+	if uid != 0 {
+		var err error
+		ctx.Data["IsApiToken"] = true
+
+		u, err = models.GetUserByID(uid)
+		if err != nil {
+			log.Error("GetUserByID:  %v", err)
+			return nil
+		}
+	}
+	token, err := models.GetAccessTokenBySHA(authToken)
+	if err == nil {
+		if isUsernameToken {
+			u, err = models.GetUserByID(token.UID)
```

Comment:
It seems that `CheckOAuthAccessToken(authToken)` checks if `authToken` is a valid JWT token and also if the granted user exists in the database. 

But even if that is the case, `models.GetAccessTokenBySHA(authToken)` checks if the last 8 symbols of the JWT token match an access token in the database. And if they do, the found user object (`u`) with `CheckOAuthAccessToken` is ignored in favour of the user object found with `GetAccessTokenBySHA`.

It would probably happen rarely (if ever) for the last 8 symbols of a JWT token to match the last 8 symbols of an access token, so the user object is unlikely to be retrieved from the database two times.

Anyway I agree it is a bit strange and the right behavior in this case seems to return the user object found by `CheckOAuthAccessToken`.

```golang
        ...
        uid := CheckOAuthAccessToken(authToken)
	if uid != 0 {
		var err error
		ctx.Data["IsApiToken"] = true

		u, err = models.GetUserByID(uid)
		if err != nil {
			log.Error("GetUserByID:  %v", err)
			return nil
		}
               return u // Return any user object found by checking the JWT user ID
	}
        token, err := models.GetAccessTokenBySHA(authToken)
        ...
```

Should I proceed and make that change?

Reply:
I'm not sure, I agree with your analysis. I'll consult with other maintainers.

Reply:
About this part of the code (the u User being potentially read two times), we'd better leave it like this. If it's in fact some kind of bug it's not your own doing, so it can be revised elsewhere. 👍